std/assembly/bindings/dom.ts

```diff
@@ -0,0 +1,162 @@
+  @external("env", "Math.cos")
+  export function cos(x: f64): f64;
+  @external("env", "Math.cosh")
+  export function cosh(x: f64): f64;
+  @external("env", "Math.exp")
+  export function exp(x: f64): f64;
+  @external("env", "Math.expm1")
+  export function expm1(x: f64): f64;
+  @external("env", "Math.floor")
+  export function floor(x: f64): f64;
+  @external("env", "Math.fround")
+  export function fround(x: f64): f32;
+  @external("env", "Math.hypot")
+  export function hypot(value1: f64, value2: f64): f64; // TODO: rest
+  @external("env", "Math.imul")
+  export function imul(a: f64, b: f64): f64;
+  @external("env", "Math.log")
+  export function log(x: f64): f64;
+  @external("env", "Math.log10")
+  export function log10(x: f64): f64;
+  @external("env", "Math.log1p")
+  export function log1p(x: f64): f64;
+  @external("env", "Math.log2")
+  export function log2(x: f64): f64;
+  @external("env", "Math.max")
+  export function max(value1: f64, value2: f64): f64; // TODO: rest
+  @external("env", "Math.min")
+  export function min(value1: f64, value2: f64): f64; // TODO: rest
+  @external("env", "Math.pow")
+  export function pow(base: f64, exponent: f64): f64;
+  @external("env", "Math.random")
+  export function random(): f64;
+  @external("env", "Math.round")
+  export function round(x: f64): f64;
+  @external("env", "Math.sign")
+  export function sign(x: f64): f64;
+  @external("env", "Math.sin")
+  export function sin(x: f64): f64;
+  @external("env", "Math.sinh")
+  export function sinh(x: f64): f64;
+  @external("env", "Math.sqrt")
+  export function sqrt(x: f64): f64;
+  @external("env", "Math.tan")
+  export function tan(x: f64): f64;
+  @external("env", "Math.tanh")
+  export function tanh(x: f64): f64;
+  @external("env", "Math.trunc")
+  export function trunc(x: f64): f64;
+}
+
+export declare namespace Reflect {
+  @external("env", "Reflect.get")
+  export function get(target: externref, propertyKey: string): externref;
+  @external("env", "Reflect.getWithReceiver")
+  @external.js("return Reflect.get(target, propertyKey, receiver);")
+  export function getWithReceiver(target: externref, propertyKey: string, receiver: externref): externref;
+  @external("env", "Reflect.has")
+  export function has(target: externref, propertyKey: string): bool;
+  @external("env", "Reflect.set")
+  export function set(target: externref, propertyKey: string, value: externref): externref;
+  @external("env", "Reflect.setWithReceiver")
+  @external.js("return Reflect.set(target, propertyKey, value, receiver);")
+  export function setWithReceiver(target: externref, propertyKey: string, value: externref , receiver: externref): externref;
+  @external("env", "Reflect.apply")
+  export function apply(target: externref, thisArgument: externref, argumentsList: externref): externref;
+}
+
+export declare namespace String {
+  @external("env", "String.fromCodePoint")
+  export function fromCodePoint(codepoint: i32): externref;
```

Comment:
So if we call this in AS, we get back an `externref`. How will it be useful?

Reply:
So far the bindings generate one half of the story, that is translating from AS -> JS. The other half, based on reference types, would be a follow-up PR. I could imagine that it would look like `declare`d classes, to annotate the capabilities of external objects. Something like

```ts
declare class JSString {
  toString(): JSString;
  ...
}
```

then "backed" by `externref`, with the bindings then switching from `externref` to `JSString` and so on. But that's still TODO and tied to eventual table allocation etc.